core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HttpHealthChecker.java

```diff
@@ -50,29 +50,55 @@
 
     private final HealthCheckerContext ctx;
     private final HttpClient httpClient;
+    private final String authority;
     private final String path;
+    private final boolean useGet;
     private boolean wasHealthy;
     private long maxLongPollingSeconds;
     @Nullable
     private HttpResponse lastResponse;
     private boolean closed;
 
-    HttpHealthChecker(HealthCheckerContext ctx, String path) {
+    HttpHealthChecker(HealthCheckerContext ctx, String path, boolean useGet) {
 
         final Endpoint endpoint = ctx.endpoint();
         final SessionProtocol protocol = ctx.protocol();
         final String scheme = protocol.uriText();
         final String ipAddr = endpoint.ipAddr();
+        final String authority = endpoint.authority();
+        final int port = ctx.port() != 0 ? ctx.port() : endpoint.port(protocol.defaultPort());
         final HttpClientBuilder builder;
         if (ipAddr == null) {
-            builder = new HttpClientBuilder(scheme + "://" + endpoint.authority());
+            this.authority = port == protocol.defaultPort() ? endpoint.host()
+                                                            : (endpoint.host() + ':' + port);
+            builder = new HttpClientBuilder(scheme + "://" + this.authority);
         } else {
-            final int port = ctx.port() > 0 ? ctx.port() : endpoint.port(protocol.defaultPort());
+            final StringBuilder uriBuf = new StringBuilder(authority.length() + 10 /* "none+https" */);
+            final StringBuilder authorityBuf = new StringBuilder(authority.length());
+            uriBuf.append(scheme);
             if (endpoint.ipFamily() == StandardProtocolFamily.INET) {
-                builder = new HttpClientBuilder(scheme + "://" + ipAddr + ':' + port);
+                uriBuf.append("://").append(ipAddr);
+                if (endpoint.isIpAddrOnly()) {
+                    authorityBuf.append(ipAddr);
+                } else {
+                    authorityBuf.append(endpoint.host());
+                }
             } else {
-                builder = new HttpClientBuilder(scheme + "://[" + ipAddr + "]:" + port);
+                uriBuf.append("://[").append(ipAddr).append(']');
+                if (endpoint.isIpAddrOnly()) {
+                    authorityBuf.append('[').append(ipAddr).append(']');
+                } else {
+                    authorityBuf.append(endpoint.host());
+                }
             }
+
+            if (port != protocol.defaultPort()) {
+                uriBuf.append(':').append(port);
+                authorityBuf.append(':').append(port);
+            }
+
+            builder = new HttpClientBuilder(uriBuf.toString());
+            this.authority = authorityBuf.toString();
```

Comment:
Instead of saving to a field and manually adding it to headers, I think we can add it as a header to the client. Is that cleaner?

Reply:
I think both are OK. I guess they don't make much difference except it will end up with making a single-header `HttpHeaders` internally, which might be tad bit more expensive.